BJS2-6262

rommstein opened the pull request:
Стать ментором

rommstein committed:
BJS2-6262
Стать ментором

rommstein committed:
BJS2-6262
Добавил дополнительные тесты

mihanchik94 commented on src/main/java/school/faang/user_service/controller/mentorship/MentorshipController.java:
```diff
@@ -0,0 +1,29 @@
+import school.faang.user_service.dto.UserDTO;
+import java.util.List;
+
+@Component
```
Лучше заменить на аннотацию @RestController.

Почему так:
Данная аннотация содержит внутри себя @Component, но также показывает то, что данный класс отвечает за обработку запросов от пользователя.

mihanchik94 commented on src/main/java/school/faang/user_service/controller/mentorship/MentorshipService.java:
```diff
@@ -0,0 +1,57 @@
+package school.faang.user_service.controller.mentorship;
```
Данный класс не относится к контроллерам и он должен располагаться в пакете service.
Примерный путь должен быть такой: school.faang.user_service.service.mentorship.MentorshipService

Почему так:
у сервисов и контроллеров разные предназначения: контроллеры отвечают за обработку запросов пользователя, а сервисы - за бизнес-логику.
И когда кому-то из коллег дадут задание дописать какую-то логику для MentorshipService, твои коллеги сразу отправятся искать этот класс в пакете service

mihanchik94 commented on src/main/java/school/faang/user_service/controller/mentorship/MentorshipService.java:
```diff
@@ -0,0 +1,57 @@
+        User mentor = mentorshipRepository.findById(userId).orElse(null);
+        if (mentor == null) {
+            return new ArrayList<>();
+        }
```
На мой взгляд, этот метод у тебя работает немного неправильно.
Тут нужно бросать исключение в случае, если пользователь не найден.

В данном случае, поскольку ниже ты почти во всех методах используешь findById(userId), ты можешь вынести этот метод отдельно и прямо в нем выкидывать исключение, если пользователь не найден.
Обычно в проектах делается отдельный пакет для исключений и в него добавляются всякие пользовательские исключения.

Почему метод работает неправильно:
Представь ситуацию со стороны пользователя приложения: он делает запрос на поиск менти по id какого-то ментора и получает пустой лист.
Т.е. пользователь приложения не поймет, что ментора с таким id нет в системе:  из работы этого метода он подумает, что этот ментор есть, но у него просто нет учеников (менти).

mihanchik94 commented on src/main/java/school/faang/user_service/controller/mentorship/MentorshipService.java:
```diff
@@ -0,0 +1,57 @@
+        User user = mentorshipRepository.findById(userId).orElse(null);
+        if (user == null) {
+            return new ArrayList<>();
+        }
```
Ситуация, аналогична методу  getMentees(long userId) (см. коммент выше).

mihanchik94 commented on src/main/java/school/faang/user_service/controller/mentorship/UserMapper.java:
```diff
@@ -0,0 +1,12 @@
+package school.faang.user_service.controller.mentorship;
```
Для данного класса тоже лучше сделать отдельный пакет, например, mapper.mentorship

mihanchik94 reviewed (commented):
После того, как все отредактируешь, не забудь и тесты поправить

rommstein committed:
BJS2-6262
Исправил методы getMentors() и getMentee() в MentorshipService
Перенёс классы в соответсвующие пакеты
Исправил аннотации
Исправил тесты

eRqa commented on src/main/java/school/faang/user_service/mapper/mentorship/UserMapper.java:
```diff
@@ -0,0 +1,12 @@
+
+@Mapper
+public interface UserMapper {
+    UserMapper INSTANCE = Mappers.getMapper(UserMapper.class);
```
Это лишнее)
Аннотация @Mapper делает этот класс бином, а ты внутри бина, подключаешь его же
Можешь посмотреть UserMapperImpl, который сгенерировал MapStruct, и INSTANCE  там не будет использоваться, он не нужен

eRqa commented on src/main/java/school/faang/user_service/service/mentorship/MentorshipService.java:
```diff
@@ -0,0 +1,54 @@
+        User mentor = mentorshipRepository.findById(userId).
+                orElseThrow(() -> new MentorshipException("Ментор с id: " + userId + " не найден"));
+        return mentor.getMentees().stream()
+                .map(UserMapper.INSTANCE::userToUserDTO)
```
Без INSTANCE))
UserMapper это уже бин, у которого можно вызвать метод userToUserDTO

eRqa commented on src/main/java/school/faang/user_service/mapper/mentorship/UserMapper.java:
```diff
@@ -0,0 +1,12 @@
+@Mapper
+public interface UserMapper {
+    UserMapper INSTANCE = Mappers.getMapper(UserMapper.class);
+    UserDTO userToUserDTO(User user);
```
Можно просто toDto

InnMish replied:
А этот метод работает? Надо в тестах проверить

eRqa commented on src/test/java/school/faang/user_service/controller/mentorship/MentorshipControllerTest.java:
```diff
@@ -0,0 +1,66 @@
+
+    @BeforeEach
+    void setUp() {
+        MockitoAnnotations.openMocks(this);
```
Для чего это добавлено?)

eRqa replied:
В других тестах тоже

eRqa commented on src/test/java/school/faang/user_service/controller/mentorship/MentorshipControllerTest.java:
```diff
@@ -0,0 +1,66 @@
+import java.util.Arrays;
+import java.util.List;
+import static org.mockito.Mockito.when;
+
```
ExtendsWith(MockitoExtension.class), без этого не будет работать @Mock & @InjectMocks

eRqa replied:
В других тестах тоже

rommstein replied:
В моём коде уже используется вариант с MockitoAnnotations.openMocks(this), который инициализирует моки перед каждым тестом.

eRqa replied:
Использование ExtendsWith более традиционный подход
К тому же с точки зрения читабельности лучше добавить одну строку, чем писать метод @BeforeEach и тело с вызовом MockitoAnnotations

eRqa commented on src/test/java/school/faang/user_service/controller/mentorship/MentorshipServiceTest.java:
```diff
@@ -0,0 +1,193 @@
+    void testGetMenteesWhenMentorNotFound() {
+        long nonExistentMentorId = 100L;
+        when(mentorshipService.getMentees(nonExistentMentorId))
+                .thenThrow(new MentorshipException("Ментор с id: " + nonExistentMentorId + " не найден"));
```
Что проверяет тест?
Цель теста - проверить совпадает ли ожидаемое поведение с фактическим, выбрасывается ли исключение когда должно и когда не должно. Здесь ты просто замокал "когда вызывается getMentees должно выбросится MentorshipException", но не проверил

InnMish commented on src/main/java/school/faang/user_service/exception/MentorshipException.java:
```diff
@@ -0,0 +1,7 @@
+package school.faang.user_service.exception;
+
+public class MentorshipException extends RuntimeException {
+    public MentorshipException(String message) {
+        super(message);
```
это слишком узкое исключение - мы его используем только в том случае, если ошибка связана с менторством 
Но если мы будет под каждую сущность создавать свое исключение, у нас будет очень много исключений с одинаковой логикой.
Я рекомендую исключения создавать по типу проблемы, а не сущности 

DataValidationException на все ошибки валидации в сервисе, EntityNotFoundException на тот случай, если искомого элемента нет в бд

InnMish commented on src/main/java/school/faang/user_service/service/mentorship/MentorshipService.java:
```diff
@@ -0,0 +1,54 @@
+    private final MentorshipRepository mentorshipRepository;
+
+    public List<UserDTO> getMentees(long userId) {
+        User mentor = mentorshipRepository.findById(userId).
+                orElseThrow(() -> new MentorshipException("Ментор с id: " + userId + " не найден"));
```
Ты этот код и ниже используешь в других методах
когда так происходит, код сразу становится кандидатом на вынесение в отдельный метод, чтобы не происходило дублирования 

Это принцип DRY - dont repeat yourself

InnMish replied:
Ментор - это пользователь 
Поэтому, чтобы получить пользователя, надо обратиться за ним в UserRepository, а не MentorshipRepository

Чтобы обратиться к UserRepository, в UserServiсe надо написать метод по доставанию пользователя из БД, а в этот класс заинжектить userService уже, чтобы обратиться у нужному методу

InnMish commented on src/main/java/school/faang/user_service/service/mentorship/MentorshipService.java:
```diff
@@ -0,0 +1,54 @@
+
+    public List<UserDTO> getMentors(long userId) {
+        User user = mentorshipRepository.findById(userId).
+                orElseThrow(() -> new MentorshipException("Пользователь с id: " + userId + " не найден"));
```
в отдельный метод и тоже обращаемся к userService, а не к mentorshipRepository

InnMish commented on src/main/java/school/faang/user_service/service/mentorship/MentorshipService.java:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    public void deleteMentee(long menteeId, long mentorId) {
+        User mentor = mentorshipRepository.findById(mentorId).orElse(null);
+        User mentee = mentorshipRepository.findById(menteeId).orElse(null);
```
Также, это действие по поиску пользователя, идем в userService  
Далее: выше у тебя есть вот такая строчка:  orElseThrow(() -> new MentorshipException("Пользователь с id: " + userId + " не найден"));  
а тут почему-то нет 
Какой смысл нам возвращать null, а потом проверять, что не null, если можно сразу кинуть ошибку, если такого пользователя у нас нет

InnMish commented on src/main/java/school/faang/user_service/service/mentorship/MentorshipService.java:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    public void deleteMentor(long menteeId, long mentorId) {
+        User mentor = mentorshipRepository.findById(mentorId).orElse(null);
+        User mentee = mentorshipRepository.findById(menteeId).orElse(null);
```
То же, что выше

rommstein committed:
BJS2-6262
Смотрели старую версию кода
Я присылал после того как пофиксил те недочёты, на которые мне указал Миша

eRqa commented:
Обрати внимание на MentorshipServiceTest:
1) Не правильно расставлены аннотации InjectMocks & Mock. Этот класс тестирует MentorshipService а не MentorshipController)) InjectMocks нужно ставить над тем, **КУДА** будут подставлены моки (заглушки), а Mock ставить над тем **какой класс нужно замокать**

2) В методе testGetMenteesWithEmptyList зачем то проверяется работа контроллера. Тест класс должен тестировать только один класс))

3) MockitoAnnotations.openMocks(this) даже идея подчеркивает что этот метод нужно использовать совместно с try-with-resources, что утяжеляет метод BeforeEach, который в свою очередь тоже избыточен ведь можно использовать ExtendsWith(...). Предлагаю все таки переделать на ExtendsWith(...)

4) метод testGetMenteesWhenMentorNotFound ничего не проверяет, он только мокает поведение mentorshipService. По первых тест должен что то проверять, делать assert чего то. Во вторых ты же тестируешь mentorshipService, а значит не нужно мокать его поведение. Мокать нужно зависимости

Вот несколько лайфхаков для написания тестов
1) Для перехода из основного класс в его тест класс есть комбинация command + shift + t или на винде вроде ctrl + shift + t. Так проще не запутаться что сейчас тестируешь
2) В идее есть очень удобная функция для разделения экрана на два. Можешь открыть в двух вкладках основной класс и тест класс, на тест класс нажать правой кнопкой и "split and move right". Идея как бы поделит пополам экран и ты сможешь видеть оба класса. Это удобно так как сразу же видно что нужно тестировать, какие моки нужно сделать
3) Тестируемый класс в тесте можно назвать sut - system under test, опять же поможет не забывать что тестируешь, а так же не парится с написанием полного имени. Просто в тест методе пишешь sut и там всегда твой тестируемый класс. Например: private MentorshipService sut; а в тест методе sut.getMentees(...). Зависимости (моки) лучше называть так же как называется класс, ну как у тебя и сделано, например private UserRepository userRepository

rommstein committed:
BJS2-6262
Снова просматривали мой старый код, до из менений всех которые я внёс

eRqa commented:
@rommstein 
1) В UserMapper убрать INSTANCE
2) UserService что за сервис?) Его никто не использует
3) MentorshipControllerTest MockitoAnnotations.openMocks(this) заменить на ExtendsWith...
4) MentorshipServiceTest подправить по комментарию https://github.com/CorporationX/user_service/pull/1012#issuecomment-2102431035
5) Решить конфликты "This branch has conflicts that must be resolved"

rommstein committed:
BJS2-6262
Заменил на @ExtendWith(MockitoExtension.class) в MentorshipControllerTest

rommstein committed:
BJS2-6262
1-Удалил UserService раз он нигде не используется
2-из UserMapper убрал INSTANCE и добавил аннотацию @Mapper(componentModel = "spring")
3- переписал все методы в MentorshipService
4-Убрал MentorshipControllerTest, т.к. сказали что он вообще не нужен
5-Переписал все тесты на методы MentorshipService

rommstein committed:
BJS2-6262
1-Удалил UserService раз он нигде не используется
2-из UserMapper убрал INSTANCE и добавил аннотацию @Mapper(componentModel = "spring")
3- переписал все методы в MentorshipService
4-Убрал MentorshipControllerTest, т.к. сказали что он вообще не нужен
5-Переписал все тесты на методы MentorshipService...

rommstein committed:
BJS2-6262
+ DataValidationException

eRqa commented on src/test/java/school/faang/user_service/controller/mentorship/MentorshipServiceTest.java:
```diff
@@ -0,0 +1,212 @@
+    public User mentee;
+
+    @Mock
+    private MentorshipRepository mentorshipRepository;
```
Это поле нужно удалить, оно не используется

eRqa commented on src/test/java/school/faang/user_service/controller/mentorship/MentorshipServiceTest.java:
```diff
@@ -0,0 +1,212 @@
+    private final Long USER_ID = 1l;
+    private final Long MENTOR_ID = 1L;
+    private final Long MENTEE_ID = 2L;
+    public User user;
```
Зачем public?)
Эти поля используются только в этом тесте, их нужно private

eRqa commented on src/main/java/school/faang/user_service/service/mentorship/MentorshipService.java:
```diff
@@ -0,0 +1,58 @@
+                .collect(Collectors.toList());
+    }
+
+    public User getUserById(long userId) {
```
Этот метод используется только в MentorshipService. Нужно сделать private)
Инкапсуляция для этого и нужна, что бы скрыть внутренне устройство, а наружу вытащить только "контракт" - публичные методы.

rommstein replied:
Понимаю, но метод в MentorshipService getUserById используется у меня в тестах testGetUserByIdWithException() и в testGetUserByIdAndRightReturnUser(), поэтому сделать его private не могу

eRqa commented on src/test/java/school/faang/user_service/controller/mentorship/MentorshipServiceTest.java:
```diff
@@ -0,0 +1,212 @@
+@ExtendWith(MockitoExtension.class)
+class MentorshipServiceTest {
+
+    private final Long USER_ID = 1l;
```
Так как это константа, нужно сделать private static final, так принято)

eRqa commented on src/main/java/school/faang/user_service/mapper/mentorship/UserMapper.java:
```diff
@@ -0,0 +1,10 @@
+import school.faang.user_service.dto.UserDTO;
+import school.faang.user_service.entity.User;
+
+@Mapper(componentModel = "spring")
```
Все что в скобках не нужно) Не знаю откуда все это берут)
Так нужно писать когда внутри маппера хочешь добавить какой нибудь бин. Например, другой маппер.
Сейчас этот функционал не востребован, в этом маппере ты не подключаешь других бинов, поэтому componentModel = "spring" не нужно
Проблема в том что когда свойство componentModel станет deprecated (это когда не рекомендуется использовать или скоро удалят), в класс UserMapper придется вносить изменения, то есть либо удалять либо искать замену. Но зачем тебе лишняя причина для изменения класса UserMapper из за функционала который не востребован? Ты создаешь зависимость от того что не нужно прямо сейчас, а может и не понадобится. Когда понадобится, другой разговор) Тогда можно и добавить
Плюс в реальной работе коллеги точно зададут вопрос, зачем делать функционал который не нужен, и свое решение нужно будет защитить

rommstein committed:
BJS2-6262
Исправил недочёты, но метод в MentorshipService getUserById используется у меня в тестах testGetUserByIdWithException() и в testGetUserByIdAndRightReturnUser(), поэтому сделать его private не могу

rommstein committed:
BJS2-6262
Исправил недочёты
